### PR TITLE
assignment1

### PR DESCRIPTION
**Pull request by ustya-k**

### CODE REVIEW

File: week1/Q1_test.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import unittest
+import Q1
+
+
+class PermutationTest(unittest.TestCase):
+
+	def testExampleCases(self):
+		self.assertEqual(True, Q1.if_permutation(Q1.normalize_word('Listen'),  Q1.normalize_word('Silent')))
```
Comment (flerdacodeu) [inline week1/Q1_test.py]: There are methods:

```python
  self.assertTrue(...)
  self.assertFalse(...)
```

that you can use here instead.

Comment (flerdacodeu) [inline week1/Q1_test.py]: There are methods:

```python
    self.assertTrue(...)
    self.assertFalse(...)
```

that you can use here instead.

File: week1/Q1_test.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import unittest
+import Q1
+
+
+class PermutationTest(unittest.TestCase):
+
+	def testExampleCases(self):
```
Comment (flerdacodeu) [inline week1/Q1_test.py:7]: Can you think of more corner cases that you might want to test against?

File: week1/Q1.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+def if_permutation(w1, w2):
+	if w1 == w2:
+		return True
+	else:
+		return False
+
+
+def main():
+	word1 = normalize_word(input())
```
Comment (flerdacodeu) [inline week1/Q1.py]: The method `input()` has an optional argument that can be used to show a prompt.
It would make it easier for someone using this code.

File: week1/Q1.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+def normalize_word(word):
+	word = word.lower()
+	word = sorted(word)
+	return word
+
+
+def if_permutation(w1, w2):
```
Comment (flerdacodeu) [inline week1/Q1.py]: The naming of this function seems a bit misleading: it does not check for permutation, and it expect the words to be already normalized. Can we update the implementation to take two arbitrary strings and return True or False whether they are a permutation of each other?

File: week1/Q1.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	if w1 == w2:
```
Comment (flerdacodeu) [inline week1/Q1.py]: Avoid:

```python
  if [...]:
    return True
  else:
    return False
```

as it is equivalent to:

```python
  return [...]
```

File: week1/Q1_test.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+		self.assertEqual(True, Q1.if_permutation(Q1.normalize_word('Listen'),  Q1.normalize_word('Silent')))
+		self.assertEqual(True, Q1.if_permutation(Q1.normalize_word('Triangle'),  Q1.normalize_word('Integral')))
+		self.assertEqual(False, Q1.if_permutation(Q1.normalize_word('Apple'),  Q1.normalize_word('Pabble')))
+
+
+	def testErrorCases(self):
+		self.assertRaises(TypeError, Q1.if_permutation, 1)
```
Comment (flerdacodeu) [inline week1/Q1_test.py]: This case is actually passing the wrong number of arguments.
What if we pass something the right number of arguments?

File: week1/Q1.py (hunk 4 above)
Comment (flerdacodeu) [inline week1/Q1.py]: Since this is your main function, it would be nice to add a comment describing what it is mean to do, what its argument are meant to be, and what the return value represents.

File: week1/Q2.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+	def clear(self):
+		self.__init__()
+
+	def add(self, x):
+		self.length += 1
+		if self.first == None:
+			self.last = self.first = Node(x, None)
+		else:
+			self.last.next = self.last = Node(x, None)
```
Comment (flerdacodeu) [inline week1/Q2.py]: While this works, I think this is rather cryptic and it depends on the evaluation order of the sub-expressions.
Sometime cool looking code is actually much harder to understand or can become buggy when modified.
I would recommend splitting this into multiple statements.

File: week1/Q2.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+class Node:
```
Comment (flerdacodeu) [inline week1/Q2.py:1]: Adding comments to the classes and methods might help. They can be brief. You do not need to comment the meaning of self.

File: week1/Q2.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class LinkedList:
+	def __init__(self):
+		self.first = None
+		self.last = None
+		self.length = 0
+
+	def clear(self):
+		self.__init__()
```
Comment (flerdacodeu) [inline week1/Q2.py]: I think it is generally beset not to call `__init__()` directly, since it is a special method in Python.
Instead, you can call `clear()` from `__init__()` instead, if you prefer.

File: week1/Q2.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+	def add(self, x):
```
Comment (flerdacodeu) [inline week1/Q2.py]: The name `x` does not tell whether this is meant to be a node or a value. `value` would work better here.

File: week1/Q2.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+		if self.first == None:
+			self.last = self.first = Node(x, None)
+		else:
+			self.last.next = self.last = Node(x, None)
+
+
+def get_kth_to_last(L, k):
+	L_len = L.length
+	if k >= L_len or k < 0:
+		return False
```
Comment (flerdacodeu) [inline week1/Q2.py]: If "False" the best value we can return here? I mean, the list may contain the value "False".

File: week1/Q2.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+	el = L.first
+	for i in range(L_len - k - 1):
+		el = el.next
+	return el.value
+
+
+def main():
+	L = LinkedList()
+	inp = input('linkedlist elements input\n')
```
Comment (flerdacodeu) [inline week1/Q2.py]: If you put a newline at the end of the input prompt, you might as well use a print statement.
For the prompt you can use something like a question mark `"? "` and have more detailed instructions before the first input.

File: week1/Q2.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+	def add(self, x):
+		self.length += 1
+		if self.first == None:
+			self.last = self.first = Node(x, None)
+		else:
+			self.last.next = self.last = Node(x, None)
+
+
+def get_kth_to_last(L, k):
```
Comment (flerdacodeu) [inline week1/Q2.py]: Usually arguments in Python are named with lowercase variables.
You can use, instead of `L`, something like `list`.

File: week1/Q2.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+def main():
+	L = LinkedList()
+	inp = input('linkedlist elements input\n')
+	while inp:
+		L.add(inp)
+		inp = input()
+	k = int(input('k input\n'))
+	
+	el = get_kth_to_last(L, k)
+	if el:
```
Comment (flerdacodeu) [inline week1/Q2.py]: What if the list contains an empty string?

File: week1/Q2_test.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+		self.assertEqual(3, Q2.get_kth_to_last(L, 6))
+ 
+	def testFalseCases(self):
+		self.assertEqual(False, Q2.get_kth_to_last(L, -1))
+		self.assertEqual(False, Q2.get_kth_to_last(L_empty, 0))
+		self.assertEqual(False, Q2.get_kth_to_last(L, 10))
+
+ 
+if __name__ == '__main__':
+	L = Q2.LinkedList()
```
Comment (flerdacodeu) [inline week1/Q2_test.py]: Creating the `L` value here is not ideal because:

* it can be modified
* it is not clear in context (when you write the test)

Instead, you can create a `setUp()` method on your test itself that can be used to store a variable as a field on the test. This variable is re-created each time, so that one test cannot affect the other.

**Review by flerdacodeu — CHANGES_REQUESTED** [review]
Thanks for your submission Ustinia. Nicely done!
Here are some suggestions about how to make it even better! Let me know if anything is unclear and we can work together on this.

File: week1/Q1_test.py (hunk 2 above)
Comment (ustya-k) [reply week1/Q1_test.py:7]: Thanks for review! I had a very important deadline, had no time to look at anything else, sorry.
Actually, I can't think of anything except empty strings. What else could've been checked?

File: week1/Q2.py (hunk 13 above)
Comment (ustya-k) [reply week1/Q2.py]: OK. But `list` is also not a very good choice while it is a data structure :)

File: week1/Q1_test.py (hunk 2 above)
Comment (flerdacodeu) [reply week1/Q1_test.py:7]: Empty string is probably the most important.

In addition, you can check things explicitly to represent some of the things that are important for the function to work correctly (which might be already coverage by some of the examples). For example:
- make sure case does not matter:
  check "abc" against "ABC"
- make sure number of occurrences does not matter:
  check "aab" matches "baa" but not "ba" or "baaa"

File: week1/Q2.py (hunk 13 above)
Comment (flerdacodeu) [reply week1/Q2.py]: You are totally right! You could use `llist` (as you did) or `lst` or `linked_list`.
I common convention is to name variables with lowercase letters and use uppercase letters for constants.

File: week1/Q2.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+# consists a value of itself and a link to the next element
+class Node:
+	def __init__(self, value = None, next = None):
+		self.value = value
+		self.next = next
+
+# data structure, consists of nodes
+# clear -- resets a linked list
+# add -- adds new node, adds a link to the new element to the last one (if present)
+class LinkedList:
```
Comment (flerdacodeu) [inline week1/Q2.py]: If the comments are formatted as docstrings (see https://www.python.org/dev/peps/pep-0257/) there are tools that will make use of them (like `help(LinkedList)` will provide access to your docstring).

**Review by flerdacodeu — APPROVED** [review]
LGTM
Thanks Ustinia for updating the code.
Just a couple of minor suggestions.
Feel free to merge the changes once you are happy with it.
